In [5]:
import math
def numAnl_11_1a(x):
    return (x*math.cos(x))-(2*x**2)+(3*x)-1

print(numAnl_11_1a(0.2), numAnl_11_1a(0.3))
print(numAnl_11_1a(1.2), numAnl_11_1a(1.3))

-0.28398668443175157 0.0066009467376817454
0.154829305372008 -0.13225152278803654


In [7]:
def numAnl_11_7a(x):
    return 1-math.e**x+(math.e-1)*math.sin((math.pi/2)*x)

print(numAnl_11_7a(0), numAnl_11_7a(1))

0.0 0.0


In [8]:
def numAnl_11_7a_der(x):
    return -math.e**x+(math.pi/2)*(math.e-1)*math.cos((math.pi/2)*x)

print(numAnl_11_7a_der(0), numAnl_11_7a_der(1))

1.6990707845418864 -2.718281828459045


# Numerical Analysis: Burden, Faires
## Chapter 2

### Examples
#### 1
Show that $f(x)=x^3+4x^2-10=0$ has a root in $[1,2]$ and use the Bisection method to determine an approximation to the root that is accurate to at least within $10^{-4}$

**Solution**:

Because $f(1)=-5$ and $f(2)=14$, the Intermediate Value Theorem ensure that this continuous function has a root in $[1,2]$.

For the first iteration of the Bisection method, we use the fact that at the midpoint of $[1,2]$, we have $f(1.5)=2.375 > 0$. This indicates we should select the interval $[1,1.5]$ for our second iteration. Then we find that $f(1.25)=-1.796875 < 0$. This indicates we should select the interval $[1.25,1.5]$ whose midpoint is $1.375$. Continuing in this manner gives us the following table.



In [8]:
def numAnl_21_3(a,b):
    x=(a+b)/2
    return x**3-7*x**2+14*x-6

# print(numAnl_21_3(3.2), numAnl_21_3(4))
numAnl_21_3(3.4,3.425)

-0.0019980468750020464

In [9]:
(3.4+3.425)/2

3.4124999999999996

### Exercise 2.1
#### 3(c)
$f(x)=x^3-7x^2+14x-6=0$
Interval: $[3.2, 4]$

$f(3.2)=-0.11$ and $f(4)=2$, the Intermediate Value Theorem ensure that this continuous function has a root in $[3.2,4]$.

The following table is generated for the bisection method:

|  $n$  |  $a_n$  |  $b_n$  |  $p_n$   |  $f(p_n)$  |
|-------|---------|---------|----------|------------|
| $1$   | $3.2$   | $4$     | $3.6$    | $0.336$    |
| $2$   | $3.2$   | $3.6$   | $3.4$    | $-0.016$   |
| $3$   | $3.4$   | $3.6$   | $3.5$    | $0.125$    |
| $4$   | $3.4$   | $3.5$   | $3.45$   | $0.046$    |
| $5$   | $3.4$   | $3.45$  | $3.425$  | $0.013$    |
| $6$   | $3.4$   | $3.425$ | $3.4125$ | $-0.002$   |

The Solution is 3.4125.

### Exercise 2.2
#### 7


In [8]:
def numAnl_22_7_fpi(x):
    return (3*x**2+3)**(1/4)

def numAnl_22_7_val(x):
    return x**4-3*x**2-3

i = 0
point = 1

In [9]:
while i<10:
    point1 = numAnl_22_7_fpi(point)
    print(f'iter:{i}, val:{round(point1,3)}, diff:{round(point1-point,3)}')
    point=point1
    i+=1

val = numAnl_22_7_val(point)
print(f'val:{round(val,5)}')

iter:0, val:1.565, diff:0.565
iter:1, val:1.794, diff:0.228
iter:2, val:1.886, diff:0.092
iter:3, val:1.923, diff:0.037
iter:4, val:1.938, diff:0.015
iter:5, val:1.943, diff:0.006
iter:6, val:1.946, diff:0.002
iter:7, val:1.947, diff:0.001
iter:8, val:1.947, diff:0.0
iter:9, val:1.947, diff:0.0
val:-0.00166


### 2.3
#### 6a

In [12]:
import numpy as np

def numAnl_23_6a_Newton(x):
    return (x - (np.exp(x) + 2**(-x) + 2 * np.cos(x) - 6)
            /(np.exp(x) - np.log(2) * 2**(-x) - 2 * np.sin(x)))

def numAnl_23_6a_val(x):
    return np.exp(x) + 2**(-x) + 2 * np.cos(x) - 6

i = 0
point = 1.5

In [13]:
while i<10:
    point1 = numAnl_23_6a_Newton(point)
    print(f'iter:{i}, val:{round(point,5)}, gap:{round(point1-point, 6)}')
    point=point1
    i+=1

val = numAnl_23_6a_val(point)
print(f'val:{round(val,3)}')

iter:0, val:1.5, gap:0.45649
iter:1, val:1.95649, gap:-0.114957
iter:2, val:1.84153, gap:-0.012027
iter:3, val:1.82951, gap:-0.000122
iter:4, val:1.82938, gap:-0.0
iter:5, val:1.82938, gap:-0.0
iter:6, val:1.82938, gap:0.0
iter:7, val:1.82938, gap:0.0
iter:8, val:1.82938, gap:0.0
iter:9, val:1.82938, gap:0.0
val:0.0


### 2.4
#### 7


### Exercise 2.1
#### 3(c)
$f(x)=x^3-7x^2+14x-6=0$
Interval: $[3.2, 4]$

$f(3.2)=-0.11$ and $f(4)=2$, the Intermediate Value Theorem ensure that this continuous function has a root in $[3.2,4]$.

The following table is generated for the bisection method:

|  $n$  |  $a_n$  |  $b_n$  |  $p_n$   |  $f(p_n)$  |
|-------|---------|---------|----------|------------|
| $1$   | $3.2$   | $4$     | $3.6$    | $0.336$    |
| $2$   | $3.2$   | $3.6$   | $3.4$    | $-0.016$   |
| $3$   | $3.4$   | $3.6$   | $3.5$    | $0.125$    |
| $4$   | $3.4$   | $3.5$   | $3.45$   | $0.046$    |
| $5$   | $3.4$   | $3.45$  | $3.425$  | $0.013$    |
| $6$   | $3.4$   | $3.425$ | $3.4125$ | $-0.002$   |

The Solution is 3.4125.

### Exercise 2.2
#### 7
Use FPI method to determine a solution accurate within $10^{-2}$ for $x^4-3x^2-3=0$ on $[1,2]$. Use $p_0=1$.

**Solution:**

$f(x)=x^4-3x^2-3=0$  
$\Rightarrow x^4=3x^2+3$  
$\Rightarrow x=(3x^2+3)^{1/4}$  

$x=g(x)=(3x^2+3)^{1/4}$  

|   | $p_n$ | $g(p_n)$ | $p_{n+1}-p_n$ |
|---|-------|----------|---------------|
| 0 | 1     | 1.565    |               |
| 1 | 1.565 | 1.794    | 0.565         |
| 2 | 1.794 | 1.886    | 0.228         |
| 3 | 1.886 | 1.923    | 0.092         |
| 4 | 1.923 | 1.938    | 0.037         |
| 5 | 1.938 | 1.943    | 0.015         |
| 6 | 1.943 | 1.946    | 0.006         |

Root=$1.943$

### 2.3
#### 6a
Use Newton's method to find solution accurate within $10^{-5}$ for: $e^x+2^{-x}+2\cos{x}-6=0$ for $1\le x\le2$

Let's take $p_0=1$

Newton's method: $$p_1=p_0-\frac{f(p_0)}{f'(p_0)}=p_0-\frac{e^{p_0}+2^{-p_0}+2\cos{p_0}-6}{e^{p_0}-\ln(2)2^{-p_0}-2\sin{p_0}}$$


$=1-\frac{f(1)}{f'(1)}=1-\frac{e^1+2^{-1}+2\cos{1}-6}{f'(1)}$

|   | $p_n$   | $p_n-\frac{f(p_n)}{f(p_n)}$ | $p_{n+1}-p_n$ |
|---|---------|-----------------------------|---------------|
| 0 | 1.5     | 1.95649                     |               |
| 1 | 1.95649 | 1.84153                     | 0.114957      |
| 2 | 1.84153 | 1.82951                     | 0.012027      |
| 3 | 1.82951 | 1.82938                     | 0.000122      |
| 4 | 1.82938 | 1.82938                     | 0.0           |

Root=$1.82938$

### 2.4
#### 7a
Show that for any positive integer k, the sequence defined by $p_n=\frac{1}{n^k}$ converges linearly to $p=0$.

**Solution:**

$\lim_{n\rightarrow\infin}{\frac{|p_{n+1}-p|}{|p_{n}-p|^{\alpha=1}}}$  
$=\lim_{n\rightarrow\infin}{\frac{|\frac{1}{(n+1)^k}-0|}{|\frac{1}{(n)^k}-0|}}$  
$=\lim_{n\rightarrow\infin}{(\frac{n}{n+1})^k}$  
$=\lim_{n\rightarrow\infin}{(\frac{n}{n(1+\frac{1}{n})})^k}$  
$=\lim_{n\rightarrow\infin}{(\frac{1}{1+\frac{1}{n}})^k}$  
$=1\gt0$  
So the sequence $p_n=\frac{1}{n^k}$ converges linearly to $p=0$.

#### 7b
For each pair of integers $k$ and $m$, determine a number $N$ for which $\frac{1}{N^k}\lt10^{-m}$

**Solution:**

$\frac{1}{N^k}\lt10^{-m}$  
$\N^k\gt10^{m}$  
$\N\gt10^{m/k}$, which is the desired number.

### 2.6
#### 1b
Find the approximations to within $10^{-4}$ to all the real zeros of the following polynomial using Netton's method.  
$f(x)=x^3+3x^2-1$

**Solution:**

Choosing $x_0=1$

| 1 | 3 | 0 | -1 |         |
|---|---|---|----|---------|
|   | 1 | 4 | 4  |         |
| 1 | 4 | 4 | 3  | $f(1)$  |
|   | 1 | 5 |    |         |
| 1 | 5 | 9 |    | $f'(1)$ |

$x_1=x_0-\frac{f(x)}{f'(x)}=1-\frac{3}{9}=0.6667$

Choosing $x_1=0.6667$

| 0.6667 | 1 | 3      | 0       | -1     |              |
|--------|---|--------|---------|--------|--------------|
|        |   | 0.6667 | 2.44444 | 1.6296 |              |
|        | 1 | 3.6667 | 2.44444 | 0.6296 | $f(0.6667)$  |
|        |   | 0.6667 | 2.8889  |        |              |
|        | 1 | 4.3333 | 5.3333  |        | $f'(0.6667)$ |

$x_2=0.6667-\frac{f(0.6667)}{f'(0.6667)}=1-\frac{0.6296}{5.3333}=0.54861$

$x_3=0.54861-\frac{f(0.54861)}{f'(0.54861)}=0.53239$

$x_4=0.53239-\frac{f(0.53239)}{f'(0.53239)}=0.53208$

$x_5=0.53208-\frac{f(0.53208)}{f'(0.53208)}=0.53208$

$|x_5-x_4|\lt10^{-4}$, $x_5=0.53208$ is a zero of the function $f(x)$.

Choosing $x_0=-1$

$x_1=-1-\frac{f(-1)}{f'(-1)}=-0.66667$

$x_2=0.66667-\frac{f(0.66667)}{f'(0.66667)}=0.65277$

$x_3=0.65277-\frac{f(0.65277)}{f'(0.65277)}=0.65270$

$|x_3-x_2|\lt10^{-4}$, $x_5=0.65270$ is a zero of the function $f(x)$.

Choosing $x_0=-3$

$x_1=-3-\frac{f(-3)}{f'(-3)}=-2.88889$

$x_2=-2.88889-\frac{f(-2.88889)}{f'(-2.88889)}=-2.87945$

$x_3=-2.87945-\frac{f(-2.87945)}{f'(-2.87945)}=-2.87938$

$|x_3-x_2|\lt10^{-4}$, $x_5=-2.87938$ is a zero of the function $f(x)$.

### 2.6
#### 1b

### 4.4

#### Exercise
##### 1(a)

In [3]:
import numpy as np

In [4]:
# Define the function
def f(x):
    return x * np.log(x)

# Given interval and number of intervals
a = 1
b = 2
n = 4  # as given in the problem

# Calculate the width of each interval
h = (b - a) / n

# Generate the x values at each point
x_values = np.linspace(a, b, n + 1)

# Apply the trapezoidal rule
integral_approx = (h / 2) * (f(x_values[0]) + 2 * sum(f(x) for x in x_values[1:-1]) + f(x_values[-1]))
integral_approx


0.639900477687986

### 4.5
#### Exercise
##### 1(a) Use Romberg integration to compute $R_{3,3}$ for the integral: $\int_{1}^{1.5}{x^2\ln{x}dx}$

In [7]:
import numpy as np

# Define the function to integrate
def f(x):
    return x**2 * np.log(x)

# Define the limits of integration
a = 1
b = 1.5

# Romberg integration with R[3,3] target
# Step 1: Initialize R matrix with trapezoidal rule approximations
R = np.zeros((4, 4))  # We need up to R[3,3]

# Compute R[k,0] using the trapezoidal rule with 2^k intervals
for k in range(4):
    n_intervals = 2 ** k
    h = (b - a) / n_intervals
    x_values = np.linspace(a, b, n_intervals + 1)
    R[k, 0] = (h / 2) * (f(x_values[0]) + 2 * sum(f(x) for x in x_values[1:-1]) + f(x_values[-1]))

# Step 2: Use Romberg extrapolation to fill in R[k,j] for j > 0
for j in range(1, 4):
    for k in range(j, 4):
        R[k, j] = (4**j * R[k, j-1] - R[k-1, j-1]) / (4**j - 1)

# Extract R[3,3]
R_3_3 = R[3, 3]
R_3_3


0.19225935770658403

In [9]:
R

array([[0.22807412, 0.        , 0.        , 0.        ],
       [0.20120251, 0.19224531, 0.        , 0.        ],
       [0.19449447, 0.19225846, 0.19225934, 0.        ],
       [0.19281809, 0.1922593 , 0.19225936, 0.19225936]])

## 1. Modeling, Computers, and Error Analysis
1.1.5 Introduction to Error Analysis
Sources of Error
- Modeling Error
- Numerical Error
    - Truncation Error
    - Roundoff Error

Accuracy Vs. Precision:
- Accuracy is how close to the bulls eye you hit.
- Precision is how narrowed or spreaded out your hits are.

True Error = True value - Approximation

Relateive error = (True Error / True Value) * 100

Approximate Error = (Current Approximation - Previous Approximation) / Approximate Value


### Mathematical Modelling and Engineering Problem Solving
### Programming and Software
### Approximations and Round-off Errors
### Truncation Errors and the Taylor Series

## 2. Roots of Equations
### Bracketing Methods
### Open Methods
### Roots of Polynomials
### Case Studies: Roots of Equations

## 3. Linear Algebraic Equations
### Gauss Elimination
### LU Decomposition and Matrix Inversion
### Special Matrices and Gauss-Sidel
### Case Studies: Linear Algebraic Equations
    
## 4. Optimization
### 1-D Unconstrained Optimization
### Multidimensional Unconstrained Optimization
### Constrained Optimization
### Case Studies: Optimization

## 5. Curve Fittinn
### Least-Squares Regression
### Interpolation
### Fourier Approximation
### Case Studies: Curve Fitting
    
## 6. Numeric Differentiation and Integration
### Newton-Cotes Integration Formulas
### Integration of Equations
### Numerical Differentiation
### Case Studies: Numerical Integration and Differentiation
    
## 7. Odrinary Differential Equations
### Runge-Kutta Methods
### Stiffness and Multistep Methods
### Boundary-Value and Eigenvalue Problems
### Case Studies: Ordinary Differential Equations
    
## 8. Partial Differential Equations
### Finite Difference: Elliptic Equations
### Finite Difference: Parabolic Equations
### Finite-Eliment Method
### Case Studies: 
